In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error, r2_score
import pandas as pd

# Load data
file_path = r'C:\Users\eoin0\OneDrive - National College of Ireland\NCI\Eoin-NCI\Semester 5\Machine Learning\Project\bmw.csv'
df = pd.read_csv(file_path)

# Convert categorical features to numeric using pd.get_dummies
df = pd.get_dummies(df, drop_first=True)

# Define features and target
X = df.drop('price', axis=1)  # features
y = df['price']  # target

# Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Scale the data (Linear Regression requires scaling)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Fit the model
model = LinearRegression()
model.fit(X_train_scaled, y_train)

# Predict and evaluate
y_pred = model.predict(X_test_scaled)

# Calculate RMSE and R²
rmse = mean_squared_error(y_test, y_pred, squared=False)
r2 = r2_score(y_test, y_pred)

print(f"RMSE: {rmse}")
print(f"R²: {r2}")


RMSE: 4257.787842417903
R²: 0.8598093284080306


C:\Users\eoin0\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\sklearn\metrics\_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


Root Mean Squared Error (RMSE): The RMSE value was found to be 4257.79, meaning that the model's predictions on car prices had an average deviation of about 4257 currency units. This indicates a reasonable degree of accuracy in predicting the car prices, with a potential error margin of this size.

R² (Coefficient of Determination): The R² value of 0.860 suggests that the model explains approximately 86% of the variation in car prices. This indicates that the model captures the main factors that influence car price well, but there may still be unexplained variance, possibly due to factors not included in the model.